src/api/services/auditTrailService.ts

```diff
@@ -1,279 +1,21 @@
-    };
-
-    return await AuditTrail.query().insert(sharingAuditTrail);
-  } catch (error) {
-    const { errorLogger } = getLoggers();
-    errorLogger.error(`Audit trails inserted failed: ${error}`, traceId);
-    throw error;
-  }
-};
-
-export const insertManageApiKeyAuditTrail = async (
-  participant: Participant,
-  userId: number,
-  userEmail: string,
-  action: AuditAction,
-  keyName: string,
-  apiRoles: string[],
+export const performAsyncOperationWithAuditTrail = async <T>(
+  auditTrail: InsertAuditTrailDTO,
   traceId: string,
-  keyId?: string,
-  newKeyName?: string,
-  newApiRoles?: string[]
-) => {
-  try {
-    const manageApiKeyTrail: Omit<AuditTrailDTO, 'id'> = {
-      userId,
-      userEmail,
-      event: AuditTrailEvents.ManageApiKey,
-      eventData: {
-        siteId: participant.siteId!,
-        action,
-        apiRoles,
-        keyName,
-        participantId: participant.id,
-        keyId,
-        newKeyName,
-        newApiRoles,
-      },
-      succeeded: false,
-    };
-
-    return await AuditTrail.query().insert(manageApiKeyTrail);
-  } catch (error) {
-    const { errorLogger } = getLoggers();
-    errorLogger.error(`Audit trails inserted failed: ${error}`, traceId);
-    throw error;
-  }
-};
-
-export const insertSharingTypesAuditTrail = async (
-  participant: Participant,
-  userId: number,
-  userEmail: string,
-  types: ClientType[],
-  traceId: string
-) => {
-  try {
-    const sharingAuditTrail: Omit<AuditTrailDTO, 'id'> = {
-      userId,
-      userEmail,
-      event: AuditTrailEvents.UpdateSharingTypes,
-      eventData: {
-        siteId: participant.siteId!,
-        allowedTypes: types,
-        participantId: participant.id,
-      },
-      succeeded: false,
-    };
-
-    return await AuditTrail.query().insert(sharingAuditTrail);
-  } catch (error) {
-    const { errorLogger } = getLoggers();
-    errorLogger.error(`Audit trails inserted failed: ${error}`, traceId);
-    throw error;
-  }
-};
-
-export const insertKeyPairAuditTrails = async (
-  participant: Participant,
-  userId: number,
-  userEmail: string,
-  action: AuditAction,
-  name: string,
-  disabled: boolean,
-  traceId: string
-) => {
-  try {
-    const keyPairAuditTrail: Omit<AuditTrailDTO, 'id'> = {
-      userId,
-      userEmail,
-      event: AuditTrailEvents.ManageKeyPair,
-      eventData: {
-        siteId: participant.siteId!,
-        action,
-        name,
-        disabled,
-        participantId: participant.id,
-      },
-      succeeded: false,
-    };
-
-    return await AuditTrail.query().insert(keyPairAuditTrail);
-  } catch (error) {
-    const { errorLogger } = getLoggers();
-    errorLogger.error(`Audit trails inserted failed: ${error}`, traceId);
-    throw error;
-  }
-};
-
-export const insertDomainNamesAuditTrails = async (
-  participant: Participant,
-  userId: number,
-  userEmail: string,
-  action: AuditAction,
-  domainNames: string[],
-  traceId: string
+  operation: () => Promise<T>
```

Comment:
hmm I don't love auditOperation or auditHandler because `operation` itself doesn't do any auditing. I could call it `auditedOperation` but I'm not sure that's an improvement. I think I'll merge as-is for now but happy to change later if we can come up with a better name.